bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java

```diff
@@ -0,0 +1,720 @@
+import org.openhab.binding.shelly.internal.coap.ShellyCoapHandler;
+import org.openhab.binding.shelly.internal.coap.ShellyCoapServer;
+import org.openhab.binding.shelly.internal.config.ShellyBindingConfiguration;
+import org.openhab.binding.shelly.internal.config.ShellyThingConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ShellyBaseHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Markus Michels - Initial contribution
+ */
+@NonNullByDefault
+public class ShellyBaseHandler extends BaseThingHandler implements ShellyDeviceListener {
+    public final Logger logger = LoggerFactory.getLogger(ShellyBaseHandler.class);
+
+    @Nullable
+    protected final NetworkAddressService networkAddressService;
+    protected final ShellyHandlerFactory handlerFactory;
+    protected ShellyThingConfiguration config = new ShellyThingConfiguration();
+    @Nullable
+    protected ShellyHttpApi api;
+    @Nullable
+    private ShellyCoapHandler coap;
+    @Nullable
+    protected ShellyDeviceProfile profile;
+    @Nullable
+    private final ShellyCoapServer coapServer;
+
+    @Nullable
+    private ScheduledFuture<?> statusJob;
+    private int skipUpdate = 0;
+    public int scheduledUpdates = 0;
+    private int skipCount = UPDATE_SKIP_COUNT;
+    private int refreshCount = UPDATE_SETTINGS_INTERVAL / UPDATE_STATUS_INTERVAL; // force settings refresh every x
+                                                                                  // seconds
+    private final int cacheCount = UPDATE_SETTINGS_INTERVAL / UPDATE_STATUS_INTERVAL; // delay before enabling channel
+                                                                                      // cache
+    private boolean refreshSettings = false;
+    private boolean channelCache = false;
+    protected boolean lockUpdates = false;
+
+    public String thingName = "";
+    private Map<String, Object> channelData = new HashMap<>();
+    protected ShellyBindingConfiguration bindingConfig = new ShellyBindingConfiguration();
+
+    /**
+     *
+     * @param thing The Thing object
+     * @param handlerFactory Handler Factory instance (will be used for event
+     *            handler registration)
+     * @param networkAddressService instance of NetworkAddressService to get access
+     *            to the OH default ip settings
+     * @param bindingConfig The binding configuration (beside thing
+     *            configuration)
+     * @param networkAddressService Network service to get local ip
+     */
+    public ShellyBaseHandler(Thing thing, ShellyHandlerFactory handlerFactory, ShellyBindingConfiguration bindingConfig,
+            @Nullable NetworkAddressService networkAddressService, @Nullable ShellyCoapServer coapServer) {
+        super(thing);
+
+        Validate.notNull(handlerFactory);
+        Validate.notNull(bindingConfig);
+        Validate.notNull(networkAddressService);
```

Comment:
> This binding uses the binding configuration for central configuration. This is not common. I would have expected a `Bridge`. What is the reason it was not implemented as a `Bridge`?

The binding configuration just holds the default user id + password. Shelly devices could be password protected, which means I need to provide credentials even to retrieve the device information. In this case the user could set a default in the binding config, which then allows to discover all devices with no protection or matching those credentials. The user could also add a thing manually and set the credentials in the thing config.

From my understanding this has nothing to do with a bridge function. I don't connect to the Shelly cloud... All things are independent, there is no gateway function.

> A note about the readme. The current version of the readme is still target towards providing a user with information when you provide the binding as a test/beta version. But that information is not relevant or should not be put in a generic readme for merged bindings. I've commented on a number of paragraphs, but do check if the readme can be improved.
Yes and no. Form my experience many users are not familiar with the basic stuff so I tried to give some general guidance how to turn on debugging, report a bug etc. and point them to the community thread. However, if that's not common sense I removed those paragraphs.

> 
> Also consider using QuantityType for channels that have a specific unit. Like Voltage.
> And it's common to mark channel advanced when you have a lot of channels and not all channels are evenly relevant.
I started to use QuantityType, but was disappointed that it doesn't defines F beside of C or percent etc. - at least if I checked the correct definitions.

Reply:
networkAddressService will be set by setNetworkAddressService after the instance is created, it's initially null and might be null if the call is not performed. So it might happen that createHandler() is called even networkAddressService has not been set.
That means

networkAddressService has to be declared as Nullable
gets passed to "new ShellyRelayHandler..." -> declared as Nullable
should not be null, but just to make sure Validate.notNull()

Reply:
You can omit this `@Nullable`, `Validate.notNull()`, and so on confusion by using _Constructor Injection_ instead of _Field Injection_ to reference the `networkAddressService` service. c'tor injection additionally allows you to add `final` to the service reference property. I just committed an update to our docs and added an example for it: https://github.com/openhab/openhab-docs/pull/1078.

Reply:
Does that mean to add
`
   @Activate
    ShellyHandlerFactory(final @Reference NetworkAddressService networkAddressService) {
        this.networkAddressService = networkAddressService;
    }
`
and remove
`
    @Reference
    protected void setNetworkAddressService(NetworkAddressService networkAddressService) {
        this.networkAddressService = networkAddressService;
    }

    protected void unsetNetworkAddressService(NetworkAddressService networkAddressService) {
        this.networkAddressService = null;
    }
`

I tried and if I did it right the new constructor is never called.

Reply:
ok, found it myself and it's done